lib/preload.js

```diff
@@ -125,9 +160,10 @@ const runContainer = (docker, container) => {
       return container.remove()
       .return(info.StatusCode)
     })
-    return container.attach({stream: true, stdout: true, stderr: true})
+    return container.attach({ stream: true, stdout: true, stderr: true, stdin: true })
     .then((stream) => {
       docker.modem.demuxStream(stream, stdout, stderr)
+      pullBuild(stdout, stream, dockerPort, imageId)
```

Comment:
This didn't changed in this PR, we return:
 * the container exit status code promise
 * the container's stdout stream
 * the container's stderr stream

Reply:
I mean that `pullBuild()` is clearly doing async stuff, that's not being waited upon, as the line after immediately returns – which seems a bit counter-intuitive to me, as this shadows error handling and such (from what I can tell).

Another thing I just noticed is that within `pullBuild()` it will dump the progress display to stdout; which is probably not what a potential consumer of this (as a module) expects (or have we moved away from publish this separately on npm?)

Reply:
If we want the caller to be able to display the stdout and stderr as it is generated, we can't wait for anything.
This no longer makes sense for stdout (as it is only used to wait for the python script) but it is still true for stderr (which is used by the python script for logging)

It's the caller's responsibility to wait for the returned `statusCodePromise`.

About the progress being written on stdout directly, you have a good point.
For now preload.js prints all progress bars to stdout (docker build, fetching image size and this new docker pull).
I'd prefer merging this PR as it is and create a new PR that stops using stdout and creates `progress` events instead as you mentioned. This would be a major change type.

Reply:
Sounds good to me!